src/prompt/variable.rs

```diff
@@ -505,6 +510,143 @@ impl VariableInput for Tuple {
     }
 }
 
+#[derive(Debug)]
+pub struct NamedTuple {
+    pub element_types: HashMap<String, Arc<dyn VariableInput>>,
+    pub shape: NamedTupleShape
+}
+
+pub struct NamedTupleParser<'a> {
+    tuple: &'a NamedTuple,
+}
+
+impl Parser<&str, Value, ParsingError> for NamedTupleParser<'_> {
+    fn parse<'a>(&mut self, mut input: &'a str) -> IResult<&'a str, Value, ParsingError> {
```

Comment:
I don't feel strongly about this, but if you want to try doing named tuples in a more "combinator-y" way, I think you could use https://docs.rs/nom/latest/nom/branch/fn.permutation.html

Reply:
Or maybe https://docs.rs/nom/latest/nom/combinator/fn.flat_map.html

Reply:
Oh, `permutation` doesn't work, for a couple reasons, the most insurmountable being that it takes a tuple, not any sort of list. I think doing a [separated_list0](https://docs.rs/nom/latest/nom/multi/fn.separated_list0.html) where the elements are [flat_map](https://docs.rs/nom/latest/nom/combinator/fn.flat_map.html)s that consume the name and map it to the right type parser could work.

The funny thing is that it might actually be easier to do a "pure combinator" approach for named tuples than for regular tuples. For regular tuples, I think we probably could do it without needing to manually implement any combinators, but I think we would need to have an explicit loop where we compose combinators. I guess that's still "pure combinator", though.